libraries/AP_Compass/AP_Compass_QMC5883P.cpp

```diff
@@ -0,0 +1,241 @@
+        delete sensor;
+        return nullptr;
+    }
+	GCS_SEND_TEXT(MAV_SEVERITY_ALERT, "sensor found");
+    return sensor;
+}
+
+AP_Compass_QMC5883P::AP_Compass_QMC5883P(AP_HAL::OwnPtr<AP_HAL::Device> dev,
+                                         bool force_external,
+                                         enum Rotation rotation)
+    : _dev(std::move(dev))
+    , _rotation(rotation)
+	, _force_external(force_external)
+{
+}
+
+bool AP_Compass_QMC5883P::init()
+{
+    _dev->get_semaphore()->take_blocking();
+
+    _dev->set_retries(10);
+#if 0
+    _dump_registers();
+#endif
+    GCS_SEND_TEXT(MAV_SEVERITY_ALERT, "Before Whoami");
+    if(!_check_whoami()){
+         GCS_SEND_TEXT(MAV_SEVERITY_ALERT, "Before Whoami Fail");
+    	 goto fail;
+    }
+
+    if (!_dev->write_register(0x0B, 0x01)||
+    	 !_dev->write_register(0x20, 0x40)||
+		 !_dev->write_register(0x21, 0x01)||
+		 !_dev->write_register(QMC5883P_REG_CONF1,
+						QMC5883P_MODE_CONTINUOUS|
+						QMC5883P_ODR_100HZ|
+						QMC5883P_OSR1_2|
+						QMC5883P_OSR2_2)||
+		 !_dev->write_register(QMC5883P_REG_CONF2,
+						0x00|
+						QMC5883P_RNG_8G|
+						0x00|
+						0x00)) {
+    		  	  goto fail;
+     }
+
+    // lower retries for run
+    _dev->set_retries(3);
+
+    _dev->get_semaphore()->give();
+
+    //register compass instance
+    _dev->set_device_type(DEVTYPE_QMC5883P);
+    if (!register_compass(_dev->get_bus_id(), _instance)) {
+        return false;
+    }
+    set_dev_id(_instance, _dev->get_bus_id());
+
+    printf("%s found on bus %u id %u address 0x%02x\n", name,
+           _dev->bus_num(), _dev->get_bus_id(), _dev->get_bus_address());
+
+    set_rotation(_instance, _rotation);
+
+    if (_force_external) {
+        set_external(_instance, true);
+    }
+
+    //Enable 100HZ
+    _dev->register_periodic_callback(10000,
+        FUNCTOR_BIND_MEMBER(&AP_Compass_QMC5883P::timer, void));
+
+    return true;
+
+ fail:
+    _dev->get_semaphore()->give();
+    return false;
+}
+bool AP_Compass_QMC5883P::_check_whoami()
+{
+    uint8_t whoami;
+    //Affected by other devices,must read registers 0x00 once or reset,after can read the ID registers reliably
+    _dev->read_registers(0x00,&whoami,1);
+    if (!_dev->read_registers(0x0C, &whoami,1)||
+      		whoami != 0x01){
+    	return false;
+    }
+    if (!_dev->read_registers(QMC5883P_REG_ID, &whoami,1)||
+    		whoami != QMC5883P_ID_VAL){
+    	return false;
+    }
+    return true;
+}
+
+void AP_Compass_QMC5883P::timer()
+{
+    struct PACKED {
+    	le16_t rx;
+    	le16_t ry;
+    	le16_t rz;
+    } buffer;
+
+    const float range_scale = 1000.0f / 3000.0f;
+
+    uint8_t status;
+    if(!_dev->read_registers(QMC5883P_REG_STATUS,&status,1)){
+    	return;
+    }
+    //new data is ready
+    if (!(status & 0x01)) {
+    	return;
+    }
+
+  if(!_dev->read_registers(QMC5883P_REG_DATA_OUTPUT_X, (uint8_t *) &buffer, sizeof(buffer))){
+	  return ;
+  }
+
+    auto x = -static_cast<int16_t>(le16toh(buffer.rx));
```

Comment:
need to explain the sign changes, or is this really a rotation?
better to use "normal" orientation for sensor, then in hwdef.dat setup rotation. Less annoying for the next vendor who uses a different rotation